opm/simulators/linalg/cuistl/PreconditionerConvertFieldTypeAdapter.hpp

```diff
@@ -0,0 +1,240 @@
+#include <opm/simulators/linalg/cuistl/detail/CuMatrixDescription.hpp>
+#include <opm/simulators/linalg/cuistl/detail/CuSparseHandle.hpp>
+#include <opm/simulators/linalg/cuistl/detail/CuSparseResource.hpp>
+#include <opm/simulators/linalg/cuistl/detail/cusparse_constants.hpp>
+#include <opm/simulators/linalg/cuistl/detail/cusparse_safe_call.hpp>
+#include <opm/simulators/linalg/cuistl/detail/preconditioner_should_call_post_pre.hpp>
+
+
+namespace Opm::cuistl
+{
+//! \brief Converts the field type (eg. double to float) to benchmark single precision preconditioners
+//!
+//! \note This is not a fast conversion, it is simply meant to benchmark the potential of some
+//!       preconditioners on consumer grade GPUs where the double precision performance is often
+//!       artificially limited.
+//!
+//! \note In theory this can do any field_type conversion that is meaningful, but it is only tested
+//! on double to float conversion.
+//!
+//! \note Remember to set the underlying preconditioner with setUnderlyingPreconditioner (should use the matrix from
+//! getConvertedMatrix())
+//!
+//! \note One could simply change the constructor design by accepting a creator function for the underlying
+//! preconditioner. For the current use cases this is however not needed.
+//!
+//! To use this, use something like the following code:
+//! \code{.cpp}
+//! #include <opm/simulators/linalg/cuistl/PreconditionerConvertFieldTypeAdapter.hpp>
+//! #include <opm/simulators/linalg/ParallelOverlappingILU0.hpp>
+//!
+//! using XDouble = Dune::BlockVector<Dune::FieldVector<double, 2>>;
+//! using MDouble = Dune::FieldMatrix<double, 2, 2>;
+//! using SpMatrixDouble = Dune::BCRSMatrix<MDouble>;
+//! using XFloat = Dune::BlockVector<Dune::FieldVector<float, 2>>;
+//! using MFloat = Dune::FieldMatrix<float, 2, 2>;
+//! using SpMatrixFloat = Dune::BCRSMatrix<MFloat>;
+//!
+//! template<class ParallelInfo>
+//! void applyILU0AsFloat(const MDouble& matrix, const XDouble& x, XDouble& y) {
+//!
+//!     using FloatILU0 = typename Opm::ParallelOverlappingILU0<MFloat, XFloat, XFloat, ParallelInfo>;
+//!     using DoubleToFloatConverter = typename Opm::cuistl::PreconditionerConvertFieldTypeAdapter<FloatILU0, MDouble,
+//!     XDouble, XDouble>;
+//!
+//!     // Note that we do not need to make a new instance for every invocation, this
+//!     // is just done for this example
+//!     auto doubleToFloatConverter = DoubleToFloatConverter(matrix);
+//!     const auto& convertedMatrix = doubleToFloatConverter.getConvertedMatrix();
+//!
+//!     auto floatILU0 = std::make_shared<FloatILU0>(convertedMatrix, 0, 1.0, 0);
+//!
+//!     doubleToFloatConverter.setUnderlyingPreconditioner(floatILU0);
+//!
+//!     // This will convert x and y to float, thenn call floatILU0.apply on the converted arguments
+//!     doubleToFloatConverter.apply(x, y);
+//! }
+//!
+//! \endcode
+template <class CudaPreconditionerType, class M, class X, class Y, int l = 1>
```

Comment:
The doc string above and the test program indicate that this class can be applied to any preconditioner. Then the template argument should be BasePreconditionerType or something like that, not using Cuda in the name. Also, it should be moved out of the cuistl subdir.

Reply:
It's  almost limited to cuistl, since  most other preconditioners use some functions that are not instantiated for float (either in OPM or Dune). It seemed non-trivial to fix. Therefore, I would I recommend keeping this as is, such that one does not expects this to work outside of cuistl.